ocaml/xenopsd/lib/xenops_sandbox.ml

```diff
@@ -104,59 +104,104 @@ end = struct
+    -> paths:Chroot.Path.t list
+    -> Chroot.t * string
+  (** [start dbg ~vm_uuid ~domid ~paths] prepares a chroot for [domid], and asks
+      the guard to create a socket restricted to [vm_uuid]. Also creates
+      empty files specified in [path] owned by [domid] user. *)
+
+  val read : domid:int -> Chroot.Path.t -> vm_uuid:string -> string
+
+  val stop : string -> domid:int -> vm_uuid:string -> unit
+  (** [strop dbg ~domid ~vm_uuid] *)
+end
+
+module type Guard = sig
+  val daemon_name : string
+
+  val create : string -> vm_uuid:Uuidm.t -> domid:int -> path:string -> unit
+
+  val destroy : string -> domid:int -> path:string -> unit
+end
+
+module MkGuard (G : Guard) : Chrooter = struct
+  let daemon = G.daemon_name
 
   let socket_path = Chroot.Path.of_string ~relative:"xapi-depriv-socket"
 
-  (** [start dbg ~vm_uuid ~domid ~paths] prepares a chroot for [domid], and asks
-      varstore-guard to create a socket restricted to [vm_uuid]. Also creates
-      empty files specified in [paths] owned by [domid] user.*)
+  let chroot ~domid ~vm_uuid = Chroot.of_domid ~daemon ~domid ~vm_uuid
+
   let start dbg ~vm_uuid ~domid ~paths =
     let chroot = Chroot.create ~daemon ~domid ~vm_uuid paths in
     let absolute_socket_path =
       Chroot.absolute_path_outside chroot socket_path
     in
-    let vm_uuidm =
+    let vm_uuid =
       match Uuidm.of_string vm_uuid with
       | Some uuid ->
           uuid
       | None ->
           failwith (Printf.sprintf "Invalid VM uuid %s" vm_uuid)
     in
-    Varstore_privileged_client.Client.create dbg vm_uuidm chroot.gid
-      absolute_socket_path ;
+    G.create dbg ~vm_uuid ~domid:chroot.gid ~path:absolute_socket_path ;
     (chroot, Chroot.chroot_path_inside socket_path)
 
-  (** [prepare ~domid path] creates an empty [path] file owned by [domid] inside
-      the chroot for [domid] and returns the absolute path to it outside the
-      chroot *)
   let prepare ~domid ~vm_uuid path =
     let chroot = Chroot.create ~daemon ~domid ~vm_uuid [path] in
     Chroot.absolute_path_outside chroot path
 
   let read ~domid path ~vm_uuid =
-    let chroot = varstored_chroot ~domid ~vm_uuid in
+    let chroot = chroot ~domid ~vm_uuid in
     path
     |> Chroot.absolute_path_outside chroot
     |> Xenops_utils.Unixext.string_of_file
 
   let stop dbg ~domid ~vm_uuid =
-    let chroot = varstored_chroot ~domid ~vm_uuid in
+    let chroot = chroot ~domid ~vm_uuid in
     if Sys.file_exists chroot.root then (
       D.debug "About to stop varstored for %d (%s) %s" domid vm_uuid __LOC__ ;
       let gid = chroot.Chroot.gid in
       let absolute_socket_path =
         Chroot.absolute_path_outside chroot socket_path
       in
       Xenops_utils.best_effort "Stop listening on deprivileged socket"
-        (fun () ->
-          Varstore_privileged_client.Client.destroy dbg gid absolute_socket_path
+        (fun () -> G.destroy dbg ~domid:gid ~path:absolute_socket_path
       ) ;
       Chroot.destroy chroot
```

Comment:
I believe only this should be guarded by `file_exists` and the rest is best effort - so let that code deal with the fact that the chroot does not exist but it still might want to do other clean-up.

Reply:
I can cherry-pick the commit in the master branch, or we can merge master branch into this one.